[Relative Load Balancing] D2 relative strategy update health score

rachelhanhan opened the pull request:
Implement the new relative load balance strategy. The overall logics are similar to the original degrader load balancer logics, we would like to call out several major logic changes:
1. In each update interval, the update is now scheduled by a single-thread executor service. However, there is one exception, if a partition is not initialized or if the uris in one partition changed, we will still update the partition state synchronously.
2. In the new implementation, each host have 1 state for each partition. It used to be one DegraderImpl object for different partitions if one host is in multiple partitions. However, in the new implementation, we changed to track each partition individually. A host can be "degraded" in one partition but in perfect health condition in another partition.
3. We used to us lowLatency as the quarantine latency, now we change to use clusterAvgLatency * lowFactor as the quarantine latency, it will be a more dynamic number.

In terms of the code structure, main logics are in RelativeStateUpdater, QuarantineManager and ClientSelectorImpl.

The unit tests will be added in next iteration.

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,217 @@
+   */
+  private void updateState()
+  {
+    // Update state for each partition
```
comment unnecessary

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,217 @@
+    } else if (_partitionLoadBalancerStateMap.get(partitionId).getClusterGenerationId() != clusterGenerationId)
+    {
+      // If there are uris change in the cluster, use executor to update the state asynchronously
+      _executorService.execute(() -> updateStateForPartition(trackerClients, partitionId, Optional.of(clusterGenerationId)));
```
Should this be setting the TrackerClients in _partitionLoadBalancerStateMap instead of passing them directly?

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,217 @@
+    ConcurrentMap<TrackerClient, TrackerClientState> trackerClientStateMap = partitionLoadBalancerState.getTrackerClientStateMap();
+
+    // Remove the trackerClients from original map if there is any change in uri list
+    if (maybeClusterGenerationId.isPresent())
```
I prefer using value directly instead of Optional

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,217 @@
+        // If it is an existing tracker client
+        double oldHealthScore = trackerClientState.getHealthScore();
+        double newHealthScore = oldHealthScore;
+        if (TrackerClientState.isUnhealthy(trackerClientState, clusterAvgLatency, callCount, latency, errorRate,
```
I'd rather keep decision making in this class instead of TrackerClientState since this is very similar to what Degrader was doing.

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,217 @@
+
+      sumAvgLatency += Math.round(latestCallStats.getCallTimeStats().getAverage());
+    }
+    long clusterAvgLatency = sumAvgLatency / trackerClients.size();
```
Calculating cluster avg latency is not this simple. See rfc for formula.

rachelhanhan replied:
Fixed

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/TrackerClientState.java:
```diff
@@ -0,0 +1,116 @@
+/**
+ * Keeps the state of each tracker client for a partition
+ */
+public class TrackerClientState {
```
maybe ServerHealthState instead?

rachelhanhan replied:
Hmm the concept of "server" is not explicitly mentioned, if the state is just tracking the state of each TrackerClient, it has a mapping relationship, I think TrackerClientState should be good

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/TrackerClientState.java:
```diff
@@ -0,0 +1,116 @@
+  // TODO: update adjusted min call count
+  private int _adjustedMinCallCount;
+  private double _healthScore;
+  private boolean _isUnhealthy;
+  private boolean _isHealthy;
```
redundant. also does this class need to know if it's healthy or not? Shouldn't the state updater calculate that and use it? Why temporarily store in this class?

rachelhanhan replied:
I changed to use an enum as you suggested during out meeting

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,184 @@
+  {
+    if (!_partitionLoadBalancerStateMap.containsKey(partitionId))
+    {
+      PartitionLoadBalancerState partitionLoadBalancerState =  new PartitionLoadBalancerState(partitionId);
```
Updated the lock logic

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,184 @@
+    {
+      PartitionLoadBalancerState partitionLoadBalancerState =  new PartitionLoadBalancerState(partitionId);
+      // If it is the very first request for the partition, update the state synchronously
+      if(partitionLoadBalancerState.getLock().tryLock())
```
Fixed

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeLoadBalancerStrategyFactory.java:
```diff
@@ -0,0 +1,182 @@
+  private static final double DEFAULT_UP_STEP = 0.05;
+  private static final double DEFAULT_DOWN_STEP = 0.2;
+  private static final double DEFAULT_RELATIVE_LATENCY_HIGH_THRESHOLD_FACTOR = 1.2;
+  private static final double DEFAULT_RELATIVE_LATENCY_LOW_THRESHOLD_FACTOR = 1.1;
```
maybe too tight. 1.4/1.3?

rachelhanhan replied:
We will decide the value after we perform more tests/simulations

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeLoadBalancerStrategyFactory.java:
```diff
@@ -0,0 +1,182 @@
+  private static final double DEFAULT_SLOW_START_THRESHOLD = 0.0;
+  private static final HttpStatusCodeRangeArray DEFAULT_ERROR_STATUS_FILTER =
+      new HttpStatusCodeRangeArray(new HttpStatusCodeRange().setLowerBound(500).setUpperBound(599));
+  private static final long DEFAULT_EMITTING_INTERVAL_MS = 0L;
```
why 0

rachelhanhan replied:
Because it was 0 in the old config too: https://jarvis.corp.linkedin.com/codesearch/result/?name=DegraderLoadBalancerStrategyConfig.java&path=rest.li%2Fd2%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fd2%2Fbalancer%2Fstrategies%2Fdegrader&reponame=linkedin%2Frest.li#136

It means by default we do not emit monitor event

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClientImpl.java:
```diff
@@ -62,6 +65,8 @@
 {
   public static final String DEFAULT_ERROR_STATUS_REGEX = "(5..)";
   public static final Pattern DEFAULT_ERROR_STATUS_PATTERN = Pattern.compile(DEFAULT_ERROR_STATUS_REGEX);
+  public static final List<HttpStatusCodeRange> DEFAULT_ERROR_STATUS_RANGES = Arrays.asList(new HttpStatusCodeRange()
```
same thing defined in Relative...Strategy

rachelhanhan replied:
Which line in RelativeLoadBalancerStrategy?

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,390 @@
+        : _partitionLoadBalancerStateMap.get(partitionId).getRing();
+  }
+
+  Map<URI, Integer> getPointsMap(int partitionId)
```
consider adding this to API so you don't need to cast to call this.

rachelhanhan replied:
This is only used by tests, the actual implementation only needs to access the ring. That's why I didn't put it to the interface

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,390 @@
+
+  private void initializePartition(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    Lock lock = new ReentrantLock();
```
concurrency concern

rachelhanhan replied:
I changed to use only one week in the StateUpdater, this will be safe, but may slow down the requests a little bit if multiple requests are sent to different partitions during initialization.

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,390 @@
+  private void updateStateForPartition(Set<TrackerClient> trackerClients, int partitionId,
+      PartitionRelativeLoadBalancerState oldPartitionState, Long clusterGenerationId)
+  {
+    debug(LOG, "Updating for partition: " + partitionId + ", state: " + oldPartitionState);
```
should we tryLock here?

zhang-chris commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,390 @@
+        || errorRate <= _relativeStrategyProperties.getLowErrorRate());
+  }
+
+  private void emitMonitorEvents(PartitionRelativeLoadBalancerState state)
```
rename to generic

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,356 @@
+  {
+    if (!isQuarantined
+        && trackerClientState.getHealthScore() == RelativeStateUpdater.MIN_HEALTH_SCORE
+        && clientWeight > MIN_ZOOKEEPER_CLIENT_WEIGHT)
```
Let's discuss about this. This line is based on the previous implementation: https://jarvis.corp.linkedin.com/codesearch/result/?name=DegraderLoadBalancerStrategyV3.java&path=rest.li%2Fd2%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinkedin%2Fd2%2Fbalancer%2Fstrategies%2Fdegrader&reponame=linkedin%2Frest.li#742

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,356 @@
+        });
+
+    // also remove the entries that the corresponding trackerClientUpdaters do not exist anymore
+    for (TrackerClient client : _healthCheckMap.keySet())
```
I think so, what would be a better place in your mind?

rachelhanhan replied:
As we discussed offline, will put the healthCheckMap to the partitionState

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/PartitionRelativeLoadBalancerState.java:
```diff
@@ -0,0 +1,239 @@
+ * Each partitionLoadBalancerState corresponds to a partition for a particular service
+ * It keeps the tracker clients and the ring for the partition
+ */
+public class PartitionRelativeLoadBalancerState
```
The name with partition will indicate this is the state for one partition, not for the entire RelativeLoadBalancer

rachelhanhan replied:
Will re-name it to RelativeLoadBalancerPartitionState.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java:
```diff
@@ -0,0 +1,141 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.d2.discovery.util.LogUtil.*;
```
Having wildcard in static import is fine in other mps' coding style, especially for voyager-api. Is there a different rule about this in pegasus?

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/PartitionLoadBalancerStateListener.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * The listener that listens to the change of a partition in relative load balancer
+ */
+public interface PartitionLoadBalancerStateListener<T>
```
@ChaoLinkedIn What do you think on this?

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,356 @@
+          quarantineMap.put(trackerClient, quarantine);
+          return true;
+        } else {
+          LOG.error("Quarantine for service {} is full! Could not add {}", _serviceName, trackerClient);
```
Will change this to warn level

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,356 @@
+        // Evict client from quarantine
+        quarantineMap.remove(quarantine);
+        quarantineHistory.put(trackerClient, quarantine);
+        LOG.info("TrackerClient {} evicted from quarantine", trackerClient.getUri());
```
As discussed in the offline, we will compare the full list of quarantine map to see if we need to log after each iteration.

ChaoLinkedIn commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/ClientSelector.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * The selector that selects a {@link TrackerClient} from the ring to route the request to
+ */
+public interface ClientSelector
```
This interface really is LoadBalancerStrategy was targeted to do. Maybe we can combine them?

rachelhanhan replied:
We have StateUpdater and ClientSelector interfaces to indicate there are 2 different steps in the load balancer strategy, so if there is one more version of strategy, we expect them to follow the same pattern to do host selection.
@zhang-chris what do you think?

ChaoLinkedIn replied:
StateUpdater is really implementation details. From user interface perspective only clientSelector matters. If you want to keep the simplicity of clientSelector, how about let LoadbalanceStrategy extends clientSelector?

rachelhanhan replied:
In that sense, I'd rather remove ClientSelector and StateUpdater as interface. I don't think strategy should extend clientSelector. If one strategy extends client selector, we should let all strategies extend client selector.
If we don't think StateUpdater will ever expose public methods to other mps, I'm fine to remove it. @zhang-chris what do you think.

rachelhanhan replied:
Removed the ClientSelector and StateUpdater interface.

ChaoLinkedIn replied:
Sorry, really did not mean to get rid of it.

Why let loadBalanceStrategy extend ClientSelector a bad idea? All existing strategies should still just need to implement LoadBalancerStrategy...

rachelhanhan replied:
Because the ClientSelector is invoked by RelativeLoadBalancerStrategy.class, client selector is used in relativeLoadBalancer to find a host from the ring after the ring is updated. StateUpdater and ClientSelector are 2 steps in the new strategy.
Because I see the interface of client selector brings the confusion, and it's more internal to the strategy implementation, so I removed the interface.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java:
```diff
@@ -0,0 +1,158 @@
+    else
+    {
+      // if this flag is set to be true, that means affinity routing is preferred but backup requests are still acceptable
+      Boolean otherHostAcceptable = KeyMapper.TargetHostHints.getRequestContextOtherHostAcceptable(requestContext);
```
Yes let's discuss about this, I saw this in the V3 implementation, so copied the exact same logic over.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/MonitorEventEmitter.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Adapter for emitting D2 events from {@link RelativeStateUpdater}.
+ */
+public class MonitorEventEmitter implements PartitionStateUpdateListener<PartitionState>
```
I didn't put Relative because it is under /relative package, do you think it's not clear?

rachelhanhan replied:
Will change to RelativeMonitorEventEmitter

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/PartitionState.java:
```diff
@@ -0,0 +1,249 @@
+    return Collections.unmodifiableList(_listeners);
+  }
+
+  public void removeTrackerClient(TrackerClient trackerClient)
```
removeTrackerClient is used in RelativeStateUpdater.java when the cluster generation id changed, we remove the hosts that de-announced

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,391 @@
+  private final double _relativeLatencyLowThresholdFactor;
+  private final RateLimitedLogger _rateLimitedLogger;
+
+  private final AtomicBoolean _quarantineEnabled;
+  private final AtomicInteger _quarantineRetries;
```
We can discuss more about this. For quarantine in particular, if we enable quarantine for one partition, we should just enable it for all partitions. I think quarantineEnabled should be a state that is shared for multiple partitions under the same service name.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,391 @@
+    }
+  }
+
+  boolean tryEnableQuarantine()
```
This is also used by unit test to explicitly enable quarantine, it encapsulates _quarantineEnabled from the direct access from unit tests.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,391 @@
+    {
+      /**
+       * Remove the client from the map if:
+       * 1. fast recovery is not enabled OR
```
I just changed the 2nd logic, I think the previous 2nd condition was actually wrong.
else if (!_fastRecoveryEnabled
        || trackerClientState.isUnhealthy()
        || trackerClientState.getHealthScore() > FAST_RECOVERY_HEALTH_SCORE_THRESHOLD)
    {
      /**
       * Remove the client from the map if:
       * 1. fast recovery is not enabled OR
       * 2. the client is still unhealthy OR
       * 3. The health score is beyond 0.5, we will let it perform normal recovery
       */
      recoverySet.remove(trackerClient);
    }

I think it is necessary to check the fast recovery, only if fast recovery is enabled, it makes sense to keep it in the recovery set.
For example, one host is quarantined, if fast recovery is not enabled, its health score can be recovered in the pattern: 0.00, 0.01, 0.06, 0.11....
if fast recovery is enabled, the health score recovery pattern can be 0.00, 0.01, 0.02, 0.04, 0.08, 0.16 if it's not getting enough traffic.

There is no point to keep a host in the recovery set if fast recovery is not enabled. Because even if it's not getting enough traffic, we cannot perform fast recovery for the host.

rachelhanhan replied:
Will remove the fast recovery check here, instead, we will add the fast recovery check when we add a new client to the recovery set.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,391 @@
+
+      if (quarantineMap.containsKey(trackerClient))
+      {
+        // If the client is still in quarantine, keep the points to 0 so no real traffic will be used
+        trackerClientState.setHealthScore(RelativeStateUpdater.MIN_HEALTH_SCORE);
+        return true;
```
I thought about this too, I think it will always be 0 as long as it is in quarantine state. I saw this logic in the V3 implementation, so I just didn't bother to reenforce the health score to be 0 again.
I can remove this reset to 0 logic, it should always be 0, and doesn't need to be reset.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java:
```diff
@@ -0,0 +1,391 @@
+    {
+      if (tryEnableQuarantine())
+      {
+        LOG.info("Quarantine is enabled for service {}", _serviceName);
```
This should be fine. This log will only happen once for each service, so it is actually information level to indicate the quarantine is now enabled.
Once _quarantineEnabled is set to true, we will not check quarantine again.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeLoadBalancerStrategyFactory.java:
```diff
@@ -0,0 +1,183 @@
+    _healthCheckOperations = healthCheckOperations;
+    _stateListenerFactories = stateListenerFactories;
+    _eventEmitter = (eventEmitter == null) ? new NoopEventEmitter() : eventEmitter;
+    _clock = SystemClock.instance();
```
The constructor of which class?

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+  void updateStateForPartition(Set<TrackerClient> trackerClients, int partitionId,
+      PartitionState oldPartitionState, Long clusterGenerationId)
+  {
+    LOG.debug("Updating for partition: " + partitionId + ", state: " + oldPartitionState);
```
I took a look into the implementation of LogUtil.debug, found it does nothing but just a check of log.isDebugEnabled(). Then I explicitly checked log4j implementation, it definitely takes care of the log level config too. So I just changed to use LOG directly since it's straightforward.

rachelhanhan replied:
Double checked on log4j implementation, it has exact same implementation of log level check and string array arguments. Will keep using log4j implementation directly.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+    for (Integer partitionId : _partitionLoadBalancerStateMap.keySet())
+    {
+      PartitionState partitionState = _partitionLoadBalancerStateMap.get(partitionId);
+      if (partitionState != null)
```
They will not, I will remove this.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+  }
+
+  /**
+   * Update the partition state when there is a cluster uris change.
```
Will update the doc

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+    // Step 3: Calculate the new ring for each partition
+    newPartitionState.resetRing();
+
+    if (clusterGenerationId != null)
```
Here is the place I used to have an optional variable. This is null if the cluster generation id does not change, it is only present if the cluster id changed. If it is changed, after all the update is done, we will update the clusterGeneratitionId of the new state.
Do you think adding more java doc will help?

rachelhanhan replied:
Will update to always pass the clusterGeneratitionId. If it's scheduled update, will pass the clusterGenerationId from the existing state instead

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+    PartitionState newPartitionState = oldPartitionState.copy();
+
+    // Step 1: Update the base health scores for each {@link TrackerClient} in the cluster
+    Map<TrackerClient, CallTracker.CallStats> latestCallStatsMap = new HashMap<>();
+    long avgClusterLatency = getAvgClusterLatency(trackerClients, latestCallStatsMap);
```
I exacted avgClusterLatency because it is also used in quarantine calculation later, if I put it into #updateHealthScoreAndState, it's hard for me to pass the value into quarantine method.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClientFactory.java:
```diff
@@ -176,18 +175,38 @@ private static Pattern getErrorStatusPattern(String loadBalancerStrategyName, Se
                                                            TransportClient transportClient,
                                                            Clock clock)
   {
+    TrackerClientImpl.ErrorStatusMatch errorStatusRangeMatch = (status) -> {
+      for(HttpStatusCodeRange statusCodeRange : getErrorStatusRanges(serviceProperties))
+      {
+        if (status >= statusCodeRange.getLowerBound() && status <= statusCodeRange.getUpperBound())
+        {
+          return true;
+        }
+      }
+      return false;
+    };
```
This is from DMRC's suggestions. I think it is easier if we want to do something such as [501 , 520] compared with using regex.

zhang-chris replied:
DMRC doesn't like using regex because it's too generic.

ChaoLinkedIn replied:
OK -- we're only compare against returncode. I'd argue generic is not a bad thing.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java:
```diff
@@ -0,0 +1,158 @@
+
+    if (targetHostUri != null)
+    {
+      trackerClient = getTrackerClientFromTarget(targetHostUri, requestContext, trackerClients);
```
I don't think so, if users are setting the fromTarget explicitly, they are expecting that host is the single host that should serve the request, if that host is not announced to d2, they should expect to get an error and fix it on their side.
This is the same logic we have in DegraderLoadBalancerStrategyV3 too.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java:
```diff
@@ -0,0 +1,158 @@
+
+    if (trackerClient == null)
+    {
+      // Pick one from the tracker clients passed from the request if the ring is completely out of date
```
This may happen a little bit more often compared with the existing V3 implementation. Because we do more async partition state update, if a request comes in and the cluster uris just changed, because we use executor to trigger the update, we may end up with a completely out-of-date ring for this request if the cluster uri set is completely new. The possibility is higher than the old implementation, but it is still very rare.

ChaoLinkedIn replied:
It's hardly to have a completely new list of URIs so agree it's rare.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/PartitionState.java:
```diff
@@ -0,0 +1,249 @@
+    resetRing();
+  }
+
+  public PartitionState copy()
```
It's not purely shallowCopy, especially for the Map variables. I created a new map for the new state, so that if we add new elements or remove outdated old elements in the new state, it will not change the old state. Want to keep the old state immutable.

ChaoLinkedIn replied:
Trying to come up a new name because copy/clone suggest reuse the objects. Maybe constructNew or build?

rachelhanhan replied:
I changed it to a constructor instead, does this work better?
public PartitionState (PartitionState oldPartitionState)
  {
    this(oldPartitionState.getPartitionId(),
        oldPartitionState.getRingFactory(),
        oldPartitionState.getPointsPerWeight(),
        new HashSet<>(oldPartitionState.getRecoveryTrackerClients()),
        oldPartitionState.getClusterGenerationId(),
        new HashMap<>(oldPartitionState.getQuarantineMap()),
        new HashMap<>(oldPartitionState.getQuarantineHistory()),
        new HashMap<>(oldPartitionState.getHealthCheckMap()),
        new HashMap<>(oldPartitionState.getTrackerClientStateMap()),
        oldPartitionState.getListeners());
  }

ChaoLinkedIn replied:
Yes!

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeLoadBalancerStrategy.java:
```diff
@@ -0,0 +1,109 @@
+  public Ring<URI> getRing(long clusterGenerationId, int partitionId, Map<URI, TrackerClient> trackerClients)
+  {
+    Ring<URI> ring = _stateUpdater.getRing(partitionId);
+    if (ring == null)
```
That's a good point. The state will keep updating itself in every interval, because it is scheduled by the executor in the constructor of the state updater. So as long as it is initialized, it will still be updated. But one case I didn't take care of is the cluster generation id change, I will just always invoke updateState then.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+            _relativeStrategyProperties.getRingProperties().getPointsPerWeight(),
+            _listenerFactories.stream().map(factory -> factory.create(partitionId)).collect(Collectors.toList()));
+
+    _lock.lock();
```
The partitionId is dynamically identified, when the RelativeStateUpdater class is initialized, we do not know how many partitions we are going to have, how can we lock by partitionId?

ChaoLinkedIn replied:
I'm thinking bitLock or lockMap. It's probably won't save us a lot given that not that many apps are using partitioning.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+    _partitionLoadBalancerStateMap = partitionLoadBalancerStateMap;
+    _lock = new ReentrantLock();
+
+    _executorService.scheduleWithFixedDelay((Runnable) this::updateState, EXECUTOR_INITIAL_DELAY,
```
My understanding is that each TrackerClient has its own interval calculation which is independent from the load balancer strategy update,  the loadBalancerStrategy will not reset the intervals in each TrackerClient. Even if we have 2 updates in the same interval, the call tracker always returns us the latest stats of the last full interval based on its own timing calculation. So I'm not worried about having insufficient data.

ChaoLinkedIn replied:
Good point -- forgot we're always reading stats from last reading window.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorImpl.java:
```diff
@@ -0,0 +1,158 @@
+    {
+      trackerClient = getTrackerClientFromRing(request, requestContext, ring, trackerClients);
+    }
+
```
Discussed offline, will apply the logic to all the cases.

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java:
```diff
@@ -0,0 +1,427 @@
+        oldPartitionState, avgClusterLatency);
+
+    // Step 3: Calculate the new ring for each partition
+    newPartitionState.resetRing();
```
Will change the naming

rachelhanhan commented on d2/src/main/java/com/linkedin/d2/balancer/clients/TrackerClientImpl.java:
```diff
@@ -280,21 +290,21 @@ private boolean isServerError(Throwable throwable)
 
-  private boolean matchErrorStatus(int status) {
-    return _errorStatusPattern.matcher(Integer.toString(status)).matches();
+  public interface ErrorStatusMatch {
+    boolean isError(int status);
   }
```
Sounds good

zhang-chris commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+            {0.25, 0.2}
+        };
+  }
+}
```
add test that verifies the executor is scheduled to run updateState periodically.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorTest.java:
```diff
@@ -0,0 +1,136 @@
+    // Ring and the tracker clients are completely off so that they do not have any overlap
+    TrackerClient trackerClient = _clientSelector.getTrackerClient(request, requestContext, DEFAULT_RING, newTrackerClientMap);
+    assertEquals(trackerClient, newTrackerClient);
+  }
```
The partially out of sync unit test is #testSubstituteClientFromRing
Are you suggesting we should always assert the picked client is added to the excluded host?

rachelhanhan replied:
Will add the test for partically excluded urls

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorTest.java:
```diff
@@ -0,0 +1,136 @@
+    KeyMapper.TargetHostHints.setRequestContextTargetHost(requestContext, URI_1);
+
+    TrackerClient trackerClient = _clientSelector.getTrackerClient(request, requestContext, DEFAULT_RING, DEFAULT_TRACKER_CLIENT_MAP);
+    assertEquals(trackerClient.getUri(), URI_1);
```
Do we normally add error message to every assert? Or just to the ones that are more flaky, so when the assert error actually happens, it tells what to check?
In most cases, we expect to pass 100% of the tests, so the error message won't show up at all.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/PartitionStateDataBuilder.java:
```diff
@@ -0,0 +1,93 @@
+/**
+ * The helper class that builds an object of {@link PartitionState}
+ */
+public class PartitionStateDataBuilder {
```
This is only for test, it is under test module already.

rachelhanhan replied:
I didn't really name it as PartitionStateBuilder because it can be confusing if one day we have an actual builder for PartitionState, the package name will be same for test data builder and non-test builder.
I can rename it to PartitionStateTestDateBuilder if that is more clear.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/PartitionStateDataBuilder.java:
```diff
@@ -0,0 +1,93 @@
+      trackerClientState.setHealthScore(healthScores.get(index));
+      trackerClientState.setHealthState(healthStates.get(index));
+      trackerClientState.setCallCount(callCountList.get(index));
+      _trackerClientStateMap.put(trackerClients.get(index), trackerClientState);
```
Need more clarification from you, not sure this is about this test method or TrackerClientState method.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,358 @@
+/*
+   Copyright (c) 2012 LinkedIn Corp.
```
Good to know that we use the current year in copyright for new files

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManagerTest.java:
```diff
@@ -0,0 +1,312 @@
+    setup(0.5, false, false);
+    _quarantineManager.tryEnableQuarantine();
+
+    List<TrackerClient> trackerClients = TrackerClientMockHelper.mockTrackerClients(2);
```
Are you asking why we only mock 2 tracker clients? To test the quarantine feature, 2 clients is enough: if we set quarantine max percentage to 50%, then we expect 1 host to be quarantined. Do you think 2 is too small?

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManagerTest.java:
```diff
@@ -0,0 +1,312 @@
+
+    _quarantineManager.updateQuarantineState(state, state, DEFAULT_AVG_CLUSTER_LATENCY);
+
+    // Verify quarantine is enabled and 2 clients are quarantined.
+    // In theory, there are 3 candidates, but because of the max percentage is 50%, we only quarantine 2
```
I think comment is for the developers to understand what we are verifying here, the assert error message is to tell why it is an error when the assert fails. I do not expect error log to give too much explanation on the purpose of the test case. For here, the error message would probably say "we should only quarantine 2 hosts"

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+    final QuarantineManager quarantineManager = Mockito.mock(QuarantineManager.class);
+
+    RelativeLoadBalancerStrategyFactory.putDefaultValues(relativeStrategyProperties);
+    Mockito.doNothing().when(quarantineManager).updateQuarantineState(any(PartitionState.class), any(PartitionState.class),
+        anyLong());
```
Will remove this.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+    // Cluster generation id changed from 0 to 1
+    _stateUpdater.updateState(new HashSet<>(trackerClients), DEFAULT_PARTITION_ID, 1);
+
+    Mockito.verify(_executorService).execute(any(Runnable.class));
```
OK will change this to wait for the execution

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+    // Verify one host's health score dropped to 80
+    assertEquals(pointsMap.get(trackerClients.get(0).getUri()).intValue(), HEALTHY_POINTS);
+    assertEquals(pointsMap.get(trackerClients.get(1).getUri()).intValue(), HEALTHY_POINTS);
+    assertEquals(pointsMap.get(trackerClients.get(2).getUri()).intValue(), 80);
```
Are you trying to get rid of using the numbers directly in unit tests? For some integration test we are going to add, we will run multiple iterations, and I just use a number to match the final health score. My point is that each test may have different scenarios and we using the calculation formula may not be straightforward sometimes. I don't see a big problem directly using numbers.

rachelhanhan replied:
I changed to use variable reference whenever I can.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+  }
+
+  @Test
+  public void testSmallCallCount()
```
I can rename it to testCallCountBelowMinCallCount

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java:
```diff
@@ -0,0 +1,359 @@
+  }
+
+  @Test
+  public void testInitializePartition()
```
I just tried to add a unit test for multiple threads initializing a partition at the same time, I found it's hard to explicitly verify the lock is only obtained once.

rachelhanhan replied:
Please take a look into the test I added and see if it works.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/ClientSelectorTest.java:
```diff
@@ -0,0 +1,136 @@
+    Mockito.when(TRACKER_CLIENT_1.getUri()).thenReturn(URI_1);
+    Mockito.when(TRACKER_CLIENT_2.getUri()).thenReturn(URI_2);
+    Mockito.when(TRACKER_CLIENT_3.getUri()).thenReturn(URI_3);
+    DEFAULT_POINTS_MAP.put(URI_1, 60);
+    DEFAULT_POINTS_MAP.put(URI_2, 80);
+    DEFAULT_POINTS_MAP.put(URI_3, 100);
+    DEFAULT_RING = new DistributionNonDiscreteRing<>(DEFAULT_POINTS_MAP);
```
For most tests I did not explicitly specify which host the ring should return. For the test that needs this, I just created a mock in that particular unit test.

rachelhanhan commented on d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManagerTest.java:
```diff
@@ -0,0 +1,312 @@
+    Mockito.when(quarantine.checkUpdateQuarantineState()).thenReturn(quarantineCheckResult);
+
+    PartitionState state = new PartitionStateDataBuilder(RING_FACTORY)
+        .setTrackerClientStateMap(trackerClients, Arrays.asList(StateUpdater.MIN_HEALTH_SCORE, 0.6, 0.6),
+            Arrays.asList(TrackerClientState.HealthState.NEUTRAL, TrackerClientState.HealthState.UNHEALTHY,
+                TrackerClientState.HealthState.UNHEALTHY),
+            Arrays.asList(20, 20, 20), RelativeLoadBalancerStrategyFactory.DEFAULT_INITIAL_HEALTH_SCORE,
+            RelativeLoadBalancerStrategyFactory.DEFAULT_MIN_CALL_COUNT)
```
I realized that if I change to array, I still have to do something like
setTrackerClientStateMap(TrackerClientMockHelper.mockTrackerClients(2),
            new double[] {StateUpdater.MIN_HEALTH_SCORE, 0.6},

I cannot directly put {A, B, C}. https://stackoverflow.com/questions/41658497/array-initializer-is-not-allowed-here

rachelhanhan commented:
Updated a few unit test and the Relative strategy converter.
We used to directly reference D2RelativeLoadBalanceStrategies in ServiceProperties, however, in the latest version we changed to use map just like all the other properties. The reasons is:
We need a map to write to and read from Zookeeper. If we directly use json converter, there can be some problems with String to long/int conversions problems. It is also more consistent to use a map instead.

Please note: The changes in ServiceProperties are backward-incompatible. ServiceProperties is mainly used in d2-config-service, even if the pegasus version is already bumped in d2-config-service, the relative strategy field is never used, so we should be good.
